Port Hug to Falcon 3.0+

vytas7 opened the pull request:
Port Hug to Falcon ~~4.0+~~ 3.0+.

Fixes #3.

vytas7 committed:
Port Hug to Falcon 4.0+.

vytas7 committed:
Check if url is not '/' before stripping the trailing slash (just in case).

jaens commented:
This looks good in general, but the `*Error(title=...)` keyword argument changes in particular look backward incompatible (if they truly are necessary)?

The possible issue is that the `hug` module directly exposes the Falcon `HTTPError` class as `hug.HTTPError`, which actual Hug apps definitely call (obviously expecting it to have the Falcon 1/2 signature).

Fortunately our own production apps always construct `*Errors` with explicit keyword arguments anyway...

vytas7 commented:
> This looks good in general, but the `*Error(title=...)` keyword argument changes in particular look backward incompatible (if they truly are necessary)?

They are truly necessary, since using non-keyword arguments for `HTTPError` was deprecated in Falcon 3.0, and removed in 4.0.
It should be backwards compatible since all Falcon versions call the parameters in question `title` and `description` (you can always use keyword arguments if you match the parameter's name).

> The possible issue is that the `hug` module directly exposes the Falcon `HTTPError` class as `hug.HTTPError`, which actual Hug apps definitely call (obviously expecting it to have the Falcon 1/2 signature).

Aha, good point, I didn't know that (not a Hug user myself)!
I'll check that, and try to wrap instead of reimporting verbatim.

**Edit**: Oh I see, it does `from falcon import *`...

jaens commented:
I think asking user to manually upgrade all occurrences of `HTTP*Error`s is also probably a valid path... anyone using hug at this point expects bit of a maintenance burden anyway.

Would be a new major version release for this repo of course....

vytas7 commented:
If you find that acceptable, just documenting it sounds the most feasible path forward... Yeah, you could call it Hug 3.0.

To be fair, I have skimmed through https://hugapi.github.io/hug/, and there doesn't seem to be (m)any references to the fact that hug is re-exporting everything from Falcon. One case in the test suite (which I have updated and haven't noticed) does use the `hug.HTTPInternalServerError` pattern, though.

Even shimming this is going to be somewhat ugly, because I would need to shadow the attributes already populated from the star (`*`) import. But maybe I can pick the most popular error classes and try to shim them back to supporting at least `title` and `description`? 

Awesome if you can check whether your production app's tests pass with my proposed changeset.

vytas7 committed:
Shim popular re-exported HTTP exceptions to accept 2 posargs again.

vytas7 commented:
@jaens I have now shimmed several popular re-exported exception classes... (Dunno if it is worth it.)

I have also added a rudimentary GitHub Actions workflow just to have minimal CI sanity-checks.
(It probably needs you to enable _Actions_ on the repository.)

vytas7 committed:
Introduce support for Falcon 3.0+. (Requires Python < 3.13.)

vytas7 commented:
@jaens @sebix I have now updated the pull request to afford Falcon 3.0+.

However, Falcon 3.x is only supported on Python < 3.13. Should we try to add complex requirements based on the Python version, or just document it?

People packaging this somewhat unlikely combination (Hug + Falcon 3.x + CPython 3.13+) for distributions could add a dependency on `legacy-cgi`.

@jaens Could you enable GitHub Actions on this repository? As mentioned before, I added some rudimentary workflows running at least `tox`.

jaens commented:
Distribution packages can _technically_ manually specify the dependencies without relying on `setup.py`, so I think just documenting should be fine.

I'm currently procrastinating trying to figure out which of my own changes to put into a _3.x_ branch and which ones to leave on the main _2.x_, but yeah, I'll try to set up Github actions :)

vytas7 commented:
> I'm currently procrastinating trying to figure out which of my own changes to put into a _3.x_ branch and which ones to leave on the main _2.x_, but yeah, I'll try to set up Github actions :)

Heh, I didn't know you were actually planning feature development, not just maintaining the bare minimum :smiling_imp:

vytas7 commented:
Not sure what changes you were planning @jaens, but it would probably make sense to cut all the Tidelift stuff, since @timothycrosley is unlikely to contract any work for a fork there anyway.

Let me know if there is anything else I can help with moving this forward.

jaens commented:
Looks good, going to test this against our production app.

(the shim is a _little_ bit risky in case someone directly checks against the `type(exc)` since now it's a subclass instead of the original type, but I suppose that's quite unlikely)

jaens commented:
Alright, merged it, still need to do some release management & testing to get it up on PyPI.

jaens commented:
Tested this in a production app, unfortunately the `Error` compatibility shims were not 100% backwards compatible.

Fortunately it was caught by tests :sweat_smile: 

The issue was eg. with this code:

```python
@hug.exception(hug.HTTPError)
def exception_handler(exception):
    LOG.debug(...)
    raise exception
```

Since `isinstance(err, hug._compat.HTTPError)` can return `False` in case the error class is *not* a shim (but the real thing), it result in these kinds of checks/handlers failing.

(I did foresee this but _thought_ it would not matter... :upside_down_face: )

vytas7 commented:
Maybe we should remove these shims altogether then, and just document the fact that Hug re-exports everything from Falcon? (It wasn't fully obvious in the current Hug docs.)

Using Falcon 3.x, the old form is still supported, so that could work as an intermediate upgrade step for certain codebases.

jaens commented:
Hmm, maybe yeah, I tried this:

```python
def _allow_title_descr_posargs(error_cls):
    class ErrorShimMeta(type):
        def __instancecheck__(cls, instance):
            return isinstance(instance, error_cls)

        def __subclasscheck__(cls, subclass):
            return issubclass(subclass, error_cls)

        def __eq__(cls, other):
            return error_cls == other

        def __hash__(cls):
            return hash(error_cls)

    class error_cls_with_posargs(error_cls, metaclass=ErrorShimMeta):
        def __new__(cls, title=None, description=None, **kwargs):
            return error_cls(title=title, description=description, **kwargs)
```

but unfortunately there is one case which it does not work for:

```python
try:
  ...
except hug.HTTPError:
  ...
```

since exception handling in Python checks the MRO directly...

(so *I think* the only option is to monkeypatch the original `Error` classes, which seems excessive)

vytas7 commented:
Hm, I'm not sure I'm following the issue wrt `HTTPError`: we are not shimming `falcon.HTTPError` / `hug.HTTPError` (they are identical).

Or do you mean that one may want to add a handler for, say, the vanilla [`falcon.HTTPGone`](https://falcon.readthedocs.io/en/stable/api/errors.html#falcon.HTTPGone), however Hug would raise a shimmed flavour somewhere that wouldn't match backwards?

Regardless of the details, I think now that I've also added back Falcon 3.0 support, the best course of action is to remove these shims, and just document it clearly that Hug re-exports certain parts of Falcon (or basically all of it). Then, in the case anyone has an extensive codebase that relies on positional `HTTPError` arguments, they can use [Falcon 3.1.3](https://falcon.readthedocs.io/en/stable/changes/3.1.3.html) as a stepping stone. Falcon 3.1.3 supports up to CPython 3.12. (And beyond with the legacy `cgi` compat package bar a couple of minor details.)

jaens commented:
The example was a bit too generalized...

It only happens for the shimmed classes, eg. if something uses `except hug.HTTPBadRequestError:` then that would not catch an exception raised via an _un-shimmed_ code path (ie. raising the original un-shimmed `falcon.HTTPBadRequestError` class).

But anyway, I think documenting the difference is the way to go.
